src/middle/SignatureMismatch.ml

```diff
@@ -0,0 +1,308 @@
+        | Some x -> wrap_func_error x
+        | None -> (
+          match check_same_type 1 return_type fun_return with
+          | Some _ ->
+              wrap_func_error
+                (ReturnTypeMismatch
+                   (ReturnType fun_return, ReturnType return_type))
+          | None ->
+              let expected_args =
+                ((UnsizedType.AutoDiffable, func_type) :: mandatory_arg_tys)
+                @ variadic_arg_tys
+              in
+              check_compatible_arguments 0 expected_args args
+              |> Option.map ~f:(fun x -> (expected_args, x)) )
+      else wrap_func_error (SuffixMismatch (FnPlain, suffix))
+  | (_, x) :: _ -> TypesMismatch (minimal_func_type, x) |> wrap_err
+  | [] -> Some ([], ArgNumMismatch (List.length mandatory_arg_tys, 0))
+
+let pp_signature_mismatch ppf (name, arg_tys, (sigs, omitted)) =
+  let open Fmt in
+  let ctx = ref TypeMap.empty in
+  let suffix_str = function
+    | Fun_kind.FnPlain -> "a pure function"
+    | FnRng -> "an rng function"
+    | FnLpdf _ -> "a probability density function"
+    | FnTarget -> "an _lp function"
+  in
+  let index_str = function
+    | 1 -> "first"
+    | 2 -> "second"
+    | 3 -> "third"
+    | 4 -> "fourth"
+    | n -> Fmt.strf "%dth" n
+  in
+  let rec pp_explain_rec ppf = function
+    | ArgError (n, DataOnlyError) ->
+        pf ppf "@[<hov>The@ %s@ argument%a@]" (index_str n) text
+          " has an incompatible data-qualifier."
+    | ArgError (n, TypesMismatch (expected, found)) ->
+        pf ppf
+          "@[<hv>The types for the %s argument are incompatible: one is@, %a@ \
+           on one but the other is@, %a@]"
+          (index_str n) (pp_unsized_type ctx) expected (pp_unsized_type ctx)
+          found
+    | ArgError (n, FuncTypeMismatch (_, _, SuffixMismatch (expected, found)))
+      ->
+        pf ppf
+          "@[<v>The %s argument is %s but the other is %s. These function \
+           types are not compatible.@]"
+          (index_str n) (suffix_str expected) (suffix_str found)
+    | ArgError (n, FuncTypeMismatch (expected, found, err)) ->
+        pf ppf
+          "@[<v>The types for the %s argument are incompatible: one is@, %a@ \
+           but the other is@, %a@ @[<v>These are not compatible because:@ \
+           @[<hov>%a@]@]@]"
+          (index_str n) (pp_fundef ctx) expected (pp_fundef ctx) found
+          pp_explain_rec err
+    | ReturnTypeMismatch (expected, found) ->
+        pf ppf "Return types are incompatible: one is %a but the other is %a"
+          UnsizedType.pp_returntype expected UnsizedType.pp_returntype found
+    | ArgNumMismatch (expected, found) ->
+        pf ppf "One takes %d arguments but the other takes %d arguments."
+          expected found
+    | SuffixMismatch (expected, found) ->
+        pf ppf
+          "One is %s but the other is %s. These function types are not \
+           compatible."
+          (suffix_str expected) (suffix_str found)
```

Comment:
That's how I originally wrote it and the toplevel `pp_explain` has "must be ... but got ..." but this `pp_explain_rec` function here is for recursing into parameter types of higher-order functions. (And _this particular branch_ only happens for higher-order functions that take higher-order functions as arguments and we don't have any of those...)
The problem is that the parameters are naturally _contravariant_. In other words, a function that expects `real` may take `int` but a function that expects `(real)=>real` cannot take `(int)=>void`, rather a function that expects `(int)=>void` may take `(real)=>void`. (I'm not sure we really want that much flexibility but that's how it was implemented before...).
This contravariance requires recursive typechecks to reverse `expected` and `found` compared to what one might assume. See the first error in the algebra_solver tests has
```
The first argument must be
 (vector, vector, data array[] real, data array[] int) => vector
but got
 (array[] real, vector, array[] real, array[] int) => vector
These are not compatible because:
  The types for the first argument are incompatible: one is
   array[] real
  on one but the other is
   vector
```
(oops, extra "on one") Note that the algebra solver expects a function that takes `vector` but the user provided one that takes `array[] real`.
I tried to make the recursive messages more ambiguous about the ordering for this reason. The toplevel part of the message identifies what was expected and what was found; the subsequent sentences only point the specific part of the types that mismatches.